syntax = "proto3";

//Enum type for different type of Genre
enum GENRE{
    FICTION = 0;
    ROMANTIC = 1;
    THRILLER = 2;
    HORROR = 3;
}    

//Numbering 1-15 is used for more frequently occuring parameters. And take 1 byte for encoding
message ISBN{
        //International Standard Book Number is a unique String assigned to each book and will be used as our primary key
        string ISBN = 1; 
}

message Book {
    string isbn = 1;

    //Stores the title of the book
    string title = 2;

    //Stores the name of the author
    string author = 3;
    int32 rating = 4;

    //Value of the year in which the book was published
    int32 publishing_year = 5; 
    GENRE genre = 6;
}

enum Status{
    taken = 0;
    available = 1;
}    

message InventoryItem {

    string Inventory_number = 9; 
    Status status = 10;

    oneof testoneof {
        Book book_object = 11;
    }
  
}

service Functionality{
    rpc getBook(ISBN) returns (Book) {};
    rpc createBook(Book) returns (ISBN) {};
}

//python3 -m grpc_tools.protoc -I. --python_out=. --pyi_out=. --grpc_python_out=. Book.proto